fish-rust/src/curses.rs

```diff
@@ -0,0 +1,300 @@
+        pub static mut cur_term: *const core::ffi::c_void;
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(
+            term: *const libc::c_char,
+            filedes: libc::c_int,
+            errret: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        /// Frees the `cur_term` TERMINAL  pointer.
+        pub fn del_curterm(term: *const core::ffi::c_void);
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`.
+        pub fn tgetflag(id: *const libc::c_char) -> libc::c_int;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+pub struct Term {
+    /// A list of cached string values corresponding to known curses string capabilities. Individual
+    /// values may be overridden with [`Term::set()`].
+    ///
+    /// The list is prepopulated on initialization in [`Term::new()`], as such the `Vec` itself is
+    /// never manipulated - only potentially the values within its individual cells.
+    strings: Vec<RefCell<Option<Arc<CString>>>>,
+}
+
+impl Term {
+    /// Looks up support for [`Capability`] `capability` in the termcap/terminfo database via the
+    /// curses library.
+    pub fn get<C: Capability>(&self, capability: C) -> C::Result {
+        capability.lookup(self)
+    }
+
+    /// Overrides the string value of `capability` for the current terminal.
+    pub fn set<S>(&self, id: StringCap, value: Option<S>)
+    where
+        S: ToCString,
+    {
+        let value = value.map(|s| Arc::new(s.to_cstring()));
+        *self.strings[id.idx()].borrow_mut() = value;
+    }
+
+    /// Initialize a new `Term` instance, prepopulating the values of all the curses string
+    /// capabilities we care about in the process.
+    fn new() -> Self {
+        let mut strings = vec![RefCell::new(None); STRING_CAPS.len()];
+        for cap in STRING_CAPS {
+            strings[cap.idx()] = cap
+                .sys_lookup()
+                // Convert the Option<CString> to a RefCell<Option<Arc<CString>>>
+                .map(Arc::new)
+                .into();
+        }
+
+        Term { strings }
+    }
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+// Clippy deduces the color of all the cows in Scotland from the first three entries we happen to
+// have here.
+#[allow(clippy::enum_variant_names)]
+enum StringCapIdx {
```

Comment:
Again it's because of how the overrides are applied at a later time, so I would need to add both a getter and a setter for each one, which I am absolutely loathe to do. And for that same reason I can't have `Option<CString>` but need the `Arc` there. 

Basically everything is complicated by the fact that the C++ code init *then* modified the term overrides rather than doing them upfront and together. As mentioned, I want to change that so this can all be cleaner, but I wanted to be faithful to the C++ code for the initial commit then change it to be prettier and faster in a separate PR.